spec/unit/http/client_spec.rb

```diff
@@ -0,0 +1,410 @@
+      expect(response.code).to eq(403)
+      expect(response.reason).to eq("Ye Shall Not Pass")
+      expect(response).to_not be_success
+    end
+
+    it 'omits basic auth if user is nil' do
+      stub_request(:get, uri).with do |req|
+        expect(req.headers).to_not include('Authorization')
+      end
+
+      client.get(uri, user: nil, password: 'pass')
+    end
+
+    it 'omits basic auth if password is nil' do
+      stub_request(:get, uri).with do |req|
+        expect(req.headers).to_not include('Authorization')
+      end
+
+      client.get(uri, user: 'user', password: nil)
+    end
+  end
+
+  context "when redirecting" do
+    let(:start_url)  { URI("https://www.example.com:8140/foo") }
+    let(:bar_url)  { "https://www.example.com:8140/bar" }
+    let(:baz_url) { "https://www.example.com:8140/baz" }
+    let(:other_host)  { "https://other.example.com:8140/qux" }
+
+    def redirect_to(status: 302, url:, body: nil)
+      { status: status, headers: { 'Location' => url }, body: body }
+    end
+
+    it "preserves GET method" do
+      stub_request(:get, start_url).to_return(redirect_to(url: bar_url))
+      stub_request(:get, bar_url).to_return(status: 200)
+
+      client.get(start_url)
+    end
+
+    it "preserves PUT method" do
+      stub_request(:put, start_url).to_return(redirect_to(url: bar_url))
+      stub_request(:put, bar_url).to_return(status: 200)
+
+      client.put(start_url, body: "", content_type: 'text/plain')
+    end
+
+    it "preserves query parameters" do
+      query = { 'debug' => true }
+      stub_request(:get, start_url).with(query: query).to_return(redirect_to(url: bar_url))
+      stub_request(:get, bar_url).with(query: query).to_return(status: 200)
+
+      client.get(start_url, params: query)
+    end
+
+    it "preserves custom and default headers when redirecting" do
+      headers = { 'X-Foo' => 'Bar', 'X-Puppet-Version' => Puppet.version }
+      stub_request(:get, start_url).with(headers: headers).to_return(redirect_to(url: bar_url))
+      stub_request(:get, bar_url).with(headers: headers).to_return(status: 200)
+
+      client.get(start_url, headers: headers)
+    end
+
+    it "redirects given a relative location" do
+      relative_url = "/people.html"
+      stub_request(:get, start_url).to_return(redirect_to(url: relative_url))
+      stub_request(:get, "https://www.example.com:8140/people.html").to_return(status: 200)
+
+      client.get(start_url)
+    end
+
+    it "preserves query parameters given a relative location" do
+      relative_url = "/people.html"
+      query = { 'debug' => true }
+      stub_request(:get, start_url).with(query: query).to_return(redirect_to(url: relative_url))
+      stub_request(:get, "https://www.example.com:8140/people.html").with(query: query).to_return(status: 200)
+
+      client.get(start_url, params: query)
+    end
+
+    [301, 302, 307].each do |code|
+      it "redirects on #{code}" do
```

Comment:
Since you use 302 as the default for the preceding tests, do you need it here?

Reply:
It's not needed. I'll update the description to say it "also redirects on"

Reply:
Updated